Fixes for io.netcdf masking

billsacks opened the pull request:
This PR includes three main changes to the application of masks (defined by _FillValue / missing_value) when io.netcdf reads data:

(1) Require that data match the _FillValue / missing_value exactly, rather than allowing approximate matches (fixes ticket #5626 )

(2) If both _FillValue and missing_value are given, give precedence to _FillValue. The netCDF standard gives special meaning to _FillValue; missing_value is just used for compatibility with old datasets.

(3) Add handling of _FillValue / missing_value of NaN

In addition, this PR adds unit tests of the reading of variables with _FillValue / missing_value attributes. The new tests cover all of the new code except for the minor refactor of __setitem__. I am assuming / hoping that this aspect of __setitem__ was covered by one or more existing tests, but that may be wrong.

billsacks commented:
Some questions:

(1) should this todo now be deleted, or are there other aspects that still need implementing?:
# * properly implement ``_FillValue``.

(I don't know the answer to this myself.)

(2) should maskandscale default to true? This would be consistent with other tools that process netcdf files, and thus would probably be the more intuitive default - though that would change behavior relative to previous versions

billsacks commented:
The discussion around issue #5626 suggested borrowing some code from netCDF4. I ended up taking very little: The _apply_missing_value routine was loosely inspired by some code from netcdf4, but more in concept than in the actual code, so I have not included the copyright information from netcdf4... let me know if you think I should. Basically, these lines were inspired by netcdf4:

```python
            try:
                missing_value_isnan = np.isnan(missing_value)
            except TypeError:
                # some data types (e.g., characters) cannot be tested for NaN
                missing_value_isnan = False
```

Some _FillValue-related features from netCDF4 that I have NOT included are:

- Taking the union of the masks defined by _FillValue and missing_value. I deliberately am not including this, because this seems like a non-standard thing to do. Really, missing_value should be ignored entirely, but I am keeping the handling of missing_value in place as a convenience to users with old datasets that define this attribute.

- Byte-swapping: I don't understand whether this is needed in io.netcdf

- Determining a default _FillValue: I don't know whether this is needed in io.netcdf

billsacks commented:
Okay, I think I'm done adding to this PR now, and am happy to have it reviewed. This is my first contribution to scipy, so I'm happy to take any suggestions.

rgommers commented on scipy/io/tests/test_netcdf.py:
```diff
@@ -45,6 +45,22 @@ def check_simple(ncfileobj):
 
+def assert_mask_matches(arr, expected_mask):
+    '''
+    Asserts that the mask of arr1 is effectively the same as expected_mask.
```
typo: ``arr1`` should be ``arr``

rgommers commented on scipy/io/tests/test_netcdf.py:
```diff
@@ -45,6 +45,22 @@ def check_simple(ncfileobj):
+    testing the 'mask' of a standard numpy array (the mask in this case is treated
+    as all False).
+
+    Arguments:
```
This is normally formatted like so:

    Parameters
    ----------
    arr: ndarray or MaskedArray
        Array to test.
    expected_mask: list of bools
        A list giving the expected mask.

rgommers replied:
I assume it's ``array_like of bools`` and not only ``list of bools`` right?

rgommers commented on scipy/io/netcdf.py:
```diff
@@ -987,6 +983,50 @@ def __setitem__(self, index, data):
+        """
+        Returns the value denoting "no data" for this variable.
+
+        If this variable does not have a missing/fill value, returns None.
```
Probably a good idea to include your comment on the PR here:

    If both _FillValue and missing_value are given, give precedence to _FillValue.
    The netCDF standard gives special meaning to _FillValue; missing_value is 
    just used for compatibility with old datasets.

rgommers replied:
Actually, I see that this reverses the decision previously made in ``__setitem__``, which gave precedence to ``missing_value``. Conceivably this can give issues for users. It may still make sense to change this, but it should be noted in the 0.18.0 release notes I think.

billsacks replied:
Yes, I deliberately reversed the logic that had been in both ```__getitem__``` and ```__setitem__``` in terms of the precedence of ```missing_value``` vs. ```_FillValue```. See my comment below for some additional rationale, in terms of making scipy.io.netcdf more consistent with other popular netCDF tools. However, I'm happy to defer to others in this regard, if there is a reason to keep ```missing_value``` as higher precedence. In practice, this will probably affect few people, because generally if a dataset has both ```_FillValue``` and ```missing_value``` they are identical. But I agree that if this is changed then it should be noted in the release notes.

rgommers commented on scipy/io/netcdf.py:
```diff
@@ -987,6 +983,50 @@ def __setitem__(self, index, data):
+                missing_value_isnan = False
+
+            if (missing_value_isnan):
+                mymask = np.isnan(data)
```
This doesn't look right. You should only use ``isnan`` if ``missing_value`` is actually ``np.nan`` right? What if you have an integer input? ``np.isnan(intarray)`` will work fine, but ``intarray`` cannot contain any nan values.

billsacks replied:
> This doesn't look right. You should only use isnan if missing_value is actually np.nan right?

Sorry, I don't understand your suggestion. Note that ```mymask = np.isnan(data)``` is already in a conditional that checks ```missing_value_isnan``` (which is set above). Is there something else you're suggesting?

>  What if you have an integer input? np.isnan(intarray) will work fine, but intarray cannot contain any nan values.

There is the possibility that a dataset will provide a missing value of NaN on integer data. To me, though, that's a mistake in the dataset (the type of the ```_FillValue``` attribute should agree with the type of the data), and I'm having trouble seeing any handling we can / should do of cases like this.

rgommers commented:
> Determining a default _FillValue: I don't know whether this is needed in io.netcdf

The code does that right? It's 999999.

rgommers commented:
> (1) should this todo now be deleted, or are there other aspects that still need implementing?:
> 
>     # * properly implement ``_FillValue``.

From the last paragraph of http://www.bic.mni.mcgill.ca/users/sean/Docs/netcdf/guide.txn_59.html:

    Sometimes there is need for more than one value to represent different kinds
    of missing data. In this case, the user should use one or more other variable
    attributes for the different kinds of missing data. For example, it might be 
    appropriate to use _FillValue to mean that data that was expected never appeared, 
    but missing_value where the creator of the data intends data to be missing, as 
    around an irregular region represented by a rectangular grid. 

So I think they're expected to be treated differently. I'm not really a netcdf user myself though, so not sure how relevant this is in practice.

> (2) should maskandscale default to true? This would be consistent with other tools that process netcdf files, and thus would probably be the more intuitive default - though that would change behavior relative to previous versions

I don't think we can change this, for backwards compatibility reasons.

rgommers commented:
> Okay, I think I'm done adding to this PR now, and am happy to have it reviewed. This is my first contribution to scipy, so I'm happy to take any suggestions.

The detailed explanation at the top of the PR and style/format of your code and commit messages is very good, all first time right.

pv commented on scipy/io/netcdf.py:
```diff
@@ -987,6 +983,50 @@ def __setitem__(self, index, data):
+        If this variable does not have a missing/fill value, returns None.
+        """
+
+        if ('_FillValue' in self._attributes):
```
Code style issue: no parenthesis in if statements
`if '_FillValue' in self._attributes:`

billsacks commented:
> (1) should this todo now be deleted, or are there other aspects that still need implementing?:
> 
> # * properly implement ``_FillValue``.
> From the last paragraph of http://www.bic.mni.mcgill.ca/users/sean/Docs/netcdf/guide.txn_59.html:
> 
> Sometimes there is need for more than one value to represent different kinds
> of missing data. In this case, the user should use one or more other variable
> attributes for the different kinds of missing data. For example, it might be 
> appropriate to use _FillValue to mean that data that was expected never appeared, 
> but missing_value where the creator of the data intends data to be missing, as 
> around an irregular region represented by a rectangular grid. 
> So I think they're expected to be treated differently. I'm not really a netcdf user myself though, so not sure how relevant this is in practice.

My read of this is that it is a suggestion for how to provide metadata documenting your data, but that does not necessarily imply that missing_value should be handled specially by generic netCDF tools like scipy.io.netcdf.

For what it's worth, I checked the handling of _FillValue / missing_value in a number of other popular netCDF utilities:

- NCO (http://nco.sourceforge.net/nco.html#Missing-Values): Treats _FillValue as special, gives no special meaning to missing_value

- NCL (https://www.ncl.ucar.edu/Document/Language/fillval.shtml): Treats _FillValue as special, gives no special meaning to missing_value

- PyNIO (checked source code): First looks for _FillValue; if not present, looks for missing_value (i.e., the same as what I have done here)

- netCDF4 python library: As noted above, takes the union of the masks defined by _FillValue and missing_value, but I feel this treatment is non-standard.

billsacks commented:
> Determining a default _FillValue: I don't know whether this is needed in io.netcdf
> The code does that right? It's 999999.

Good point. netCDF4 has more sophisticated handling of this, providing different defaults depending on the data type. I don't know how important this is, but I noted it in terms of remaining masking-related features present in netCDF4 but not here.

billsacks commented:
@rgommers and @pv : Thank you for your detailed review of this change! I have addressed all of your comments either in replies or in an update to the branch that I just pushed.

pv commented:
Thanks, rebased and merged in 43d8eb0e9678990a6b218c6a97153c355fa3dffc
